File: src/main/scala/ff/compiler/Compiler.scala
package ff.compiler
import ff.compiler.Emitter_._

import ff.compiler.Inference_._

import ff.compiler.Parser_._

import ff.compiler.Resolver_._

import ff.compiler.Syntax_._

import ff.compiler.Tokenizer_._

import ff.core.Array_._

import ff.core.ArrayBuilder_._

import ff.core.Bool_._

import ff.core.Char_._

import ff.core.Core_._

import ff.core.FileSystem_._

import ff.core.Int_._

import ff.core.List_._

import ff.core.Log_._

import ff.core.Map_._

import ff.core.Nothing_._

import ff.core.Option_._

import ff.core.Pair_._

import ff.core.Set_._

import ff.core.String_._

import ff.core.System_._

import ff.core.Try_._

import ff.core.Unit_._
object Compiler_ {

case class Compiler(files_ : ff.core.FileSystem_.FileSystem, outputPath_ : ff.core.String_.String, packagePaths_ : ff.core.Map_.Map[ff.core.String_.String, ff.core.String_.String], var parsedModules_ : ff.core.Map_.Map[ff.core.String_.String, ff.compiler.Syntax_.Module], var resolvedModules_ : ff.core.Map_.Map[ff.core.String_.String, ff.compiler.Syntax_.Module], var inferredModules_ : ff.core.Map_.Map[ff.core.String_.String, ff.compiler.Syntax_.Module], var emittedModules_ : ff.core.Set_.Set[ff.core.String_.String])
val coreImports_ : ff.core.List_.List[ff.compiler.Syntax_.DImport] = ff.core.List_.List_map[ff.core.String_.String, ff.compiler.Syntax_.DImport](self_ = List("Array", "ArrayBuilder", "Bool", "Char", "Core", "FileSystem", "Int", "List", "Log", "Map", "Nothing", "Option", "Pair", "Set", "String", "System", "Try", "Unit"), body_ = {(moduleName_) =>
ff.compiler.Syntax_.DImport(at_ = ff.compiler.Syntax_.Location(file_ = "<prelude>", line_ = 1, column_ = 1), alias_ = moduleName_, package_ = ff.core.Pair_.Pair[ff.core.String_.String, ff.core.String_.String](first_ = "ff", second_ = "core"), directory_ = List(), file_ = moduleName_)
})
def make_(files_ : ff.core.FileSystem_.FileSystem, outputPath_ : ff.core.String_.String, packagePaths_ : ff.core.Map_.Map[ff.core.String_.String, ff.core.String_.String]) : ff.compiler.Compiler_.Compiler = {
ff.compiler.Compiler_.Compiler(files_ = files_, outputPath_ = outputPath_, packagePaths_ = packagePaths_, parsedModules_ = ff.core.Map_.empty_[ff.core.String_.String, ff.compiler.Syntax_.Module](), resolvedModules_ = ff.core.Map_.empty_[ff.core.String_.String, ff.compiler.Syntax_.Module](), inferredModules_ = ff.core.Map_.empty_[ff.core.String_.String, ff.compiler.Syntax_.Module](), emittedModules_ = ff.core.Set_.empty_[ff.core.String_.String]())
}
def Compiler_parse(self_ : ff.compiler.Compiler_.Compiler, packageName_ : ff.core.String_.String, moduleName_ : ff.core.String_.String) : ff.compiler.Syntax_.Module = (self_, packageName_, moduleName_) match {
case (self_, _, _) =>
ff.core.Option_.Option_else(self_ = ff.core.Map_.Map_get[ff.core.String_.String, ff.compiler.Syntax_.Module](self_ = self_.parsedModules_, key_ = ((packageName_ + ":") + moduleName_)), body_ = {() =>
val packagePair_ : ff.core.Pair_.Pair[ff.core.String_.String, ff.core.String_.String] = ff.core.Core_.do_[ff.core.Pair_.Pair[ff.core.String_.String, ff.core.String_.String]](body_ = {() =>
val array_ : ff.core.Array_.Array[ff.core.String_.String] = ff.core.String_.String_split(self_ = packageName_, char_ = ':');
ff.core.Pair_.Pair[ff.core.String_.String, ff.core.String_.String](first_ = ff.core.Array_.Array_expect[ff.core.String_.String](self_ = array_, index_ = 0), second_ = ff.core.Array_.Array_expect[ff.core.String_.String](self_ = array_, index_ = 1))
});
val packagePath_ : ff.core.String_.String = ff.core.Map_.Map_expect[ff.core.String_.String, ff.core.String_.String](self_ = self_.packagePaths_, key_ = packageName_);
val file_ : ff.core.String_.String = (moduleName_ + ".ff");
val code_ : ff.core.String_.String = ff.core.FileSystem_.FileSystem_readText(self_ = self_.files_, file_ = ((packagePath_ + "/") + file_));
val tokens_ : ff.core.Array_.Array[ff.compiler.Token_.Token] = ff.compiler.Tokenizer_.tokenize_(file_ = file_, code_ = code_);
val module_ : ff.compiler.Syntax_.Module = ff.compiler.Parser_.Parser_parseModule(self_ = ff.compiler.Parser_.make_(packagePair_ = packagePair_, file_ = file_, tokens_ = tokens_));
val result_ : ff.compiler.Syntax_.Module = pipe_dot(module_)({(_c) =>
ff.compiler.Syntax_.Module(packagePair_ = _c.packagePair_, file_ = _c.file_, dependencies_ = _c.dependencies_, imports_ = (ff.compiler.Compiler_.coreImports_ ++ module_.imports_), types_ = _c.types_, traits_ = _c.traits_, instances_ = _c.instances_, extends_ = _c.extends_, lets_ = _c.lets_, functions_ = _c.functions_)
});
self_.parsedModules_ = ff.core.Map_.Map_add[ff.core.String_.String, ff.compiler.Syntax_.Module](self_ = self_.parsedModules_, key_ = ((packageName_ + ":") + moduleName_), value_ = result_);
result_
})
}

def Compiler_imports(self_ : ff.compiler.Compiler_.Compiler, packageName_ : ff.core.String_.String, module_ : ff.compiler.Syntax_.Module) : ff.core.List_.List[ff.compiler.Syntax_.Module] = (self_, packageName_, module_) match {
case (self_, _, _) =>
ff.core.List_.List_map[ff.compiler.Syntax_.DImport, ff.compiler.Syntax_.Module](self_ = module_.imports_, body_ = {(import_) =>
val newPackageName_ : ff.core.String_.String = ((import_.package_.first_ + ":") + import_.package_.second_);
val newModuleName_ : ff.core.String_.String = (ff.core.List_.List_join(self_ = ff.core.List_.List_map[ff.core.String_.String, ff.core.String_.String](self_ = import_.directory_, body_ = {(_w1) =>
(_w1 + "/")
}), separator_ = "") + import_.file_);
ff.compiler.Compiler_.Compiler_parse(self_ = self_, packageName_ = newPackageName_, moduleName_ = newModuleName_)
})
}

def Compiler_resolve(self_ : ff.compiler.Compiler_.Compiler, packageName_ : ff.core.String_.String, moduleName_ : ff.core.String_.String) : ff.compiler.Syntax_.Module = (self_, packageName_, moduleName_) match {
case (self_, _, _) =>
ff.core.Option_.Option_else(self_ = ff.core.Map_.Map_get[ff.core.String_.String, ff.compiler.Syntax_.Module](self_ = self_.resolvedModules_, key_ = ((packageName_ + ":") + moduleName_)), body_ = {() =>
val module_ : ff.compiler.Syntax_.Module = ff.compiler.Compiler_.Compiler_parse(self_ = self_, packageName_ = packageName_, moduleName_ = moduleName_);
val otherModules_ : ff.core.List_.List[ff.compiler.Syntax_.Module] = ff.compiler.Compiler_.Compiler_imports(self_ = self_, packageName_ = packageName_, module_ = module_);
val result_ : ff.compiler.Syntax_.Module = ff.compiler.Resolver_.Resolver_resolveModule(self_ = ff.compiler.Resolver_.make_(), module_ = module_, otherModules_ = otherModules_);
self_.resolvedModules_ = ff.core.Map_.Map_add[ff.core.String_.String, ff.compiler.Syntax_.Module](self_ = self_.resolvedModules_, key_ = ((packageName_ + ":") + moduleName_), value_ = result_);
result_
})
}

def Compiler_infer(self_ : ff.compiler.Compiler_.Compiler, packageName_ : ff.core.String_.String, moduleName_ : ff.core.String_.String) : ff.compiler.Syntax_.Module = (self_, packageName_, moduleName_) match {
case (self_, _, _) =>
ff.core.Option_.Option_else(self_ = ff.core.Map_.Map_get[ff.core.String_.String, ff.compiler.Syntax_.Module](self_ = self_.inferredModules_, key_ = ((packageName_ + ":") + moduleName_)), body_ = {() =>
val module_ : ff.compiler.Syntax_.Module = ff.compiler.Compiler_.Compiler_resolve(self_ = self_, packageName_ = packageName_, moduleName_ = moduleName_);
val otherModules_ : ff.core.List_.List[ff.compiler.Syntax_.Module] = ff.core.List_.List_map[ff.compiler.Syntax_.Module, ff.compiler.Syntax_.Module](self_ = ff.compiler.Compiler_.Compiler_imports(self_ = self_, packageName_ = packageName_, module_ = module_), body_ = {(i_) =>
val newPackageName_ : ff.core.String_.String = ((i_.packagePair_.first_ + ":") + i_.packagePair_.second_);
ff.compiler.Compiler_.Compiler_resolve(self_ = self_, packageName_ = newPackageName_, moduleName_ = ff.core.FileSystem_.FileSystem_prefixName(self_ = self_.files_, path_ = i_.file_))
});
val instances_ : ff.core.List_.List[ff.compiler.Syntax_.DInstance] = (module_.instances_ ++ ff.core.List_.List_flatMap[ff.compiler.Syntax_.Module, ff.compiler.Syntax_.DInstance](self_ = otherModules_, body_ = {(_w1) =>
_w1.instances_
}));
val result_ : ff.compiler.Syntax_.Module = ff.compiler.Inference_.Inference_inferModule(self_ = ff.compiler.Inference_.make_(instances_ = instances_), module_ = module_, otherModules_ = otherModules_);
self_.inferredModules_ = ff.core.Map_.Map_add[ff.core.String_.String, ff.compiler.Syntax_.Module](self_ = self_.inferredModules_, key_ = ((packageName_ + ":") + moduleName_), value_ = result_);
result_
})
}

def Compiler_emit(self_ : ff.compiler.Compiler_.Compiler, packageName_ : ff.core.String_.String, moduleName_ : ff.core.String_.String) : ff.core.Unit_.Unit = (self_, packageName_, moduleName_) match {
case (self_, _, _) =>
ff.core.Option_.Option_else(self_ = ff.core.Core_.if_[ff.core.Unit_.Unit](condition_ = ff.core.Set_.Set_contains[ff.core.String_.String](self_ = self_.emittedModules_, value_ = ((packageName_ + ":") + moduleName_)), body_ = {() =>
ff.core.Unit_.Unit();
ff.core.Unit_.Unit()
}), body_ = {() =>
self_.emittedModules_ = ff.core.Set_.Set_add[ff.core.String_.String](self_ = self_.emittedModules_, value_ = ((packageName_ + ":") + moduleName_));
val module_ : ff.compiler.Syntax_.Module = ff.compiler.Compiler_.Compiler_infer(self_ = self_, packageName_ = packageName_, moduleName_ = moduleName_);
ff.core.List_.List_each[ff.compiler.Syntax_.Module](self_ = ff.compiler.Compiler_.Compiler_imports(self_ = self_, packageName_ = packageName_, module_ = module_), body_ = {(i_) =>
val newPackageName_ : ff.core.String_.String = ((i_.packagePair_.first_ + ":") + i_.packagePair_.second_);
ff.compiler.Compiler_.Compiler_emit(self_ = self_, packageName_ = newPackageName_, moduleName_ = ff.core.FileSystem_.FileSystem_prefixName(self_ = self_.files_, path_ = i_.file_));
ff.core.Unit_.Unit()
});
val packagePair_ : ff.core.Pair_.Pair[ff.core.String_.String, ff.core.String_.String] = ff.core.Core_.do_[ff.core.Pair_.Pair[ff.core.String_.String, ff.core.String_.String]](body_ = {() =>
val array_ : ff.core.Array_.Array[ff.core.String_.String] = ff.core.String_.String_split(self_ = packageName_, char_ = ':');
ff.core.Pair_.Pair[ff.core.String_.String, ff.core.String_.String](first_ = ff.core.Array_.Array_expect[ff.core.String_.String](self_ = array_, index_ = 0), second_ = ff.core.Array_.Array_expect[ff.core.String_.String](self_ = array_, index_ = 1))
});
val emitted_ : ff.core.String_.String = ff.compiler.Emitter_.emitModule_(packagePair_ = packagePair_, module_ = module_);
val path_ : ff.core.String_.String = ((self_.outputPath_ + "/") + ff.core.String_.String_replace(self_ = packageName_, needle_ = ":", replacement_ = "/"));
val file_ : ff.core.String_.String = (((path_ + "/") + moduleName_) + ".scala");
ff.core.FileSystem_.FileSystem_createDirectories(self_ = self_.files_, path_ = path_);
ff.core.FileSystem_.FileSystem_writeText(self_ = self_.files_, file_ = file_, text_ = emitted_);
ff.core.Unit_.Unit()
});
ff.core.Unit_.Unit()
}


}

File: src/main/js/ff/compiler/Compiler.mjs
import * as ff_compiler_Compiler from "../../ff/compiler/Compiler.mjs"

import * as ff_compiler_Inference from "../../ff/compiler/Inference.mjs"

import * as ff_compiler_JsEmitter from "../../ff/compiler/JsEmitter.mjs"

import * as ff_compiler_Parser from "../../ff/compiler/Parser.mjs"

import * as ff_compiler_Resolver from "../../ff/compiler/Resolver.mjs"

import * as ff_compiler_Syntax from "../../ff/compiler/Syntax.mjs"

import * as ff_compiler_Tokenizer from "../../ff/compiler/Tokenizer.mjs"

import * as ff_core_Array from "../../ff/core/Array.mjs"

import * as ff_core_ArrayBuilder from "../../ff/core/ArrayBuilder.mjs"

import * as ff_core_Bool from "../../ff/core/Bool.mjs"

import * as ff_core_Char from "../../ff/core/Char.mjs"

import * as ff_core_Core from "../../ff/core/Core.mjs"

import * as ff_core_Duration from "../../ff/core/Duration.mjs"

import * as ff_core_FileSystem from "../../ff/core/FileSystem.mjs"

import * as ff_core_Float from "../../ff/core/Float.mjs"

import * as ff_core_Int from "../../ff/core/Int.mjs"

import * as ff_core_List from "../../ff/core/List.mjs"

import * as ff_core_Log from "../../ff/core/Log.mjs"

import * as ff_core_Map from "../../ff/core/Map.mjs"

import * as ff_core_Nothing from "../../ff/core/Nothing.mjs"

import * as ff_core_Option from "../../ff/core/Option.mjs"

import * as ff_core_Pair from "../../ff/core/Pair.mjs"

import * as ff_core_Set from "../../ff/core/Set.mjs"

import * as ff_core_String from "../../ff/core/String.mjs"

import * as ff_core_System from "../../ff/core/System.mjs"

import * as ff_core_TimeSystem from "../../ff/core/TimeSystem.mjs"

import * as ff_core_Try from "../../ff/core/Try.mjs"

import * as ff_core_Unit from "../../ff/core/Unit.mjs"

// type Compiler
export function Compiler(files_, scalaOutputPath_, jsOutputPath_, packagePaths_, parsedModules_, resolvedModules_, inferredModules_, emittedModules_) {
return {files_, scalaOutputPath_, jsOutputPath_, packagePaths_, parsedModules_, resolvedModules_, inferredModules_, emittedModules_};
}

export const coreImports_ = ff_core_List.List_map(ff_core_List.Link("Array", ff_core_List.Link("ArrayBuilder", ff_core_List.Link("Bool", ff_core_List.Link("Char", ff_core_List.Link("Core", ff_core_List.Link("Duration", ff_core_List.Link("FileSystem", ff_core_List.Link("Float", ff_core_List.Link("Int", ff_core_List.Link("List", ff_core_List.Link("Log", ff_core_List.Link("Map", ff_core_List.Link("Nothing", ff_core_List.Link("Option", ff_core_List.Link("Pair", ff_core_List.Link("Set", ff_core_List.Link("String", ff_core_List.Link("System", ff_core_List.Link("TimeSystem", ff_core_List.Link("Try", ff_core_List.Link("Unit", ff_core_List.Empty()))))))))))))))))))))), ((moduleName_) => {
return ff_compiler_Syntax.DImport(ff_compiler_Syntax.Location("<prelude>", 1, 1), moduleName_, ff_core_Pair.Pair("ff", "core"), ff_core_List.Empty(), moduleName_)
}))

export function make_(files_, scalaOutputPath_, jsOutputPath_, packagePaths_) {
return ff_compiler_Compiler.Compiler(files_, scalaOutputPath_, jsOutputPath_, packagePaths_, ff_core_Map.empty_(), ff_core_Map.empty_(), ff_core_Map.empty_(), ff_core_Set.empty_())
}

export function Compiler_parse(self_, packageName_, moduleName_) {
return ff_core_Option.Option_else(ff_core_Map.Map_get(self_.parsedModules_, ((packageName_ + ":") + moduleName_)), (() => {
const packagePair_ = ff_core_Core.do_((() => {
const array_ = ff_core_String.String_split(packageName_, 58)
return ff_core_Pair.Pair(ff_core_Array.Array_expect(array_, 0), ff_core_Array.Array_expect(array_, 1))
}))
const packagePath_ = ff_core_Map.Map_expect(self_.packagePaths_, packageName_)
const file_ = (moduleName_ + ".ff")
const code_ = ff_core_FileSystem.FileSystem_readText(self_.files_, ((packagePath_ + "/") + file_))
const tokens_ = ff_compiler_Tokenizer.tokenize_(file_, code_)
const module_ = ff_compiler_Parser.Parser_parseModule(ff_compiler_Parser.make_(packagePair_, file_, tokens_))
const result_ = (((_c) => {
return ff_compiler_Syntax.Module(_c.packagePair_, _c.file_, _c.dependencies_, ff_core_List.List_addAll(ff_compiler_Compiler.coreImports_, module_.imports_), _c.types_, _c.traits_, _c.instances_, _c.extends_, _c.lets_, _c.functions_)
}))(module_)
self_.parsedModules_ = ff_core_Map.Map_add(self_.parsedModules_, ((packageName_ + ":") + moduleName_), result_)
return result_
}))
}

export function Compiler_imports(self_, packageName_, module_) {
return ff_core_List.List_map(module_.imports_, ((import_) => {
const newPackageName_ = ((import_.package_.first_ + ":") + import_.package_.second_)
const newModuleName_ = (ff_core_List.List_join(ff_core_List.List_map(import_.directory_, ((_w1) => {
return (_w1 + "/")
})), "") + import_.file_)
return ff_compiler_Compiler.Compiler_parse(self_, newPackageName_, newModuleName_)
}))
}

export function Compiler_resolve(self_, packageName_, moduleName_) {
return ff_core_Option.Option_else(ff_core_Map.Map_get(self_.resolvedModules_, ((packageName_ + ":") + moduleName_)), (() => {
const module_ = ff_compiler_Compiler.Compiler_parse(self_, packageName_, moduleName_)
const otherModules_ = ff_compiler_Compiler.Compiler_imports(self_, packageName_, module_)
const result_ = ff_compiler_Resolver.Resolver_resolveModule(ff_compiler_Resolver.make_(), module_, otherModules_)
self_.resolvedModules_ = ff_core_Map.Map_add(self_.resolvedModules_, ((packageName_ + ":") + moduleName_), result_)
return result_
}))
}

export function Compiler_infer(self_, packageName_, moduleName_) {
return ff_core_Option.Option_else(ff_core_Map.Map_get(self_.inferredModules_, ((packageName_ + ":") + moduleName_)), (() => {
const module_ = ff_compiler_Compiler.Compiler_resolve(self_, packageName_, moduleName_)
const otherModules_ = ff_core_List.List_map(ff_compiler_Compiler.Compiler_imports(self_, packageName_, module_), ((i_) => {
const newPackageName_ = ((i_.packagePair_.first_ + ":") + i_.packagePair_.second_)
return ff_compiler_Compiler.Compiler_resolve(self_, newPackageName_, ff_core_FileSystem.FileSystem_prefixName(self_.files_, i_.file_))
}))
const instances_ = ff_core_List.List_addAll(module_.instances_, ff_core_List.List_flatMap(otherModules_, ((_w1) => {
return _w1.instances_
})))
const result_ = ff_compiler_Inference.Inference_inferModule(ff_compiler_Inference.make_(instances_), module_, otherModules_)
self_.inferredModules_ = ff_core_Map.Map_add(self_.inferredModules_, ((packageName_ + ":") + moduleName_), result_)
return result_
}))
}

export function Compiler_emit(self_, packageName_, moduleName_) {
if(ff_core_Set.Set_contains(self_.emittedModules_, ((packageName_ + ":") + moduleName_))) {

} else {
self_.emittedModules_ = ff_core_Set.Set_add(self_.emittedModules_, ((packageName_ + ":") + moduleName_))
const module_ = ff_compiler_Compiler.Compiler_infer(self_, packageName_, moduleName_)
const otherModules_ = ff_core_List.List_map(ff_compiler_Compiler.Compiler_imports(self_, packageName_, module_), ((i_) => {
const newPackageName_ = ((i_.packagePair_.first_ + ":") + i_.packagePair_.second_)
const newModuleName_ = ff_core_FileSystem.FileSystem_prefixName(self_.files_, i_.file_)
ff_compiler_Compiler.Compiler_emit(self_, newPackageName_, newModuleName_)
return ff_compiler_Compiler.Compiler_infer(self_, newPackageName_, newModuleName_)
}))
const packagePair_ = ff_core_Core.do_((() => {
const array_ = ff_core_String.String_split(packageName_, 58)
return ff_core_Pair.Pair(ff_core_Array.Array_expect(array_, 0), ff_core_Array.Array_expect(array_, 1))
}))
const js_ = ff_compiler_JsEmitter.JsEmitter_emitModule(ff_compiler_JsEmitter.make_(ff_core_List.Link(module_, otherModules_)), packagePair_, module_)
const jsPath_ = ((self_.jsOutputPath_ + "/") + ff_core_String.String_replace(packageName_, ":", "/"))
const jsFile_ = (((jsPath_ + "/") + moduleName_) + ".mjs")
ff_core_FileSystem.FileSystem_createDirectories(self_.files_, jsPath_)
ff_core_FileSystem.FileSystem_writeText(self_.files_, jsFile_, js_)
}
}


